### ISSUE TITLE
crash and refleaks when calling sqlite3.Cursor.__init__() more than once

### ISSUE THREAD

**Issue by 4e331a12-c2bb-4cd1-b83c-dcc766e8fbf8** [python/cpython#75951]
BPO | [31770](https://bugs.python.org/issue31770)
--- | :---
Nosy | @vstinner, @orenmn
PRs | <li>python/cpython#3968</li><li>python/cpython#4301</li><li>python/cpython#4302</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-11-07.00:50:12.995>
created_at = <Date 2017-10-12.08:18:36.858>
labels = ['extension-modules', '3.7', 'type-crash']
title = 'crash and refleaks when calling sqlite3.Cursor.__init__() more than once'
updated_at = <Date 2017-11-07.00:50:12.994>
user = 'https://github.com/orenmn'
```

bugs.python.org fields:
```python
activity = <Date 2017-11-07.00:50:12.994>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-11-07.00:50:12.995>
closer = 'vstinner'
components = ['Extension Modules']
creation = <Date 2017-10-12.08:18:36.858>
creator = 'Oren Milman'
dependencies = []
files = []
hgrepos = []
issue_num = 31770
keywords = ['patch']
message_count = 5.0
messages = ['304220', '305692', '305698', '305699', '305700']
nosy_count = 2.0
nosy_names = ['vstinner', 'Oren Milman']
pr_nums = ['3968', '4301', '4302']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'crash'
url = 'https://bugs.python.org/issue31770'
versions = ['Python 2.7', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by 4e331a12-c2bb-4cd1-b83c-dcc766e8fbf8** [comment]
The following code crashes:
import sqlite3
import weakref
def callback(*args):
    pass

```py
connection = sqlite3.connect(":memory:")
cursor = sqlite3.Cursor(connection)
ref = weakref.ref(cursor, callback)
cursor.__init__(connection)
del cursor
del ref
```


IIUC, this is because pysqlite_cursor_init() (in [Modules/_sqlite/cursor.c](https://github.com/python/cpython/blob/main/Modules/_sqlite/cursor.c)) sets
`self->in_weakreflist` to NULL, and thus corrupts the weakref list. Later,
clear_weakref() (in [Objects/weakrefobject.c](https://github.com/python/cpython/blob/main/Objects/weakrefobject.c)) tries to remove a reference from
the corrupted list, and crashes.

In every other place (that i saw) where such a weakreflist field is used, it is
set to NULL right after allocating the object (usually in __new__()), or just
not set at all, e.g. in `functools.partial`.
So since PyType_GenericNew() is the __new__() of sqlite3.Cursor, ISTM that the
simplest solution is to not touch `self->in_weakreflist` at all in
pysqlite_cursor_init().


Also, the following code results in refleaks:
import sys
import sqlite3
connection = sqlite3.connect(":memory:")
cursor = sqlite3.Cursor(connection)
refcount_before = sys.gettotalrefcount()
cursor.\_\_init__(connection)
print(sys.gettotalrefcount() - refcount_before)  # should be close to 0

This is because pysqlite_cursor_init() doesn't decref before assigning to
fields of `self`.


I would open a PR to fix this soon.

**Comment by vstinner** [comment]
New changeset e56ab746a965277ffcc4396d8a0902b6e072d049 by Victor Stinner (Oren Milman) in branch 'master':
bpo-31770: Prevent a crash and refleaks when calling sqlite3.Cursor.\_\_init__() more than once (bpo-3968)
https://github.com/python/cpython/commit/e56ab746a965277ffcc4396d8a0902b6e072d049

**Comment by vstinner** [comment]
New changeset 4b544aadd54fd2337d2ab5c137389cae8d1fd781 by Victor Stinner (Miss Islington (bot)) in branch '2.7':
bpo-31770: Prevent a crash and refleaks when calling sqlite3.Cursor.\_\_init__() more than once (GH-3968) (bpo-4302)
https://github.com/python/cpython/commit/4b544aadd54fd2337d2ab5c137389cae8d1fd781

**Comment by vstinner** [comment]
New changeset 9684cf69e32ae442c7be54521073ac78557f3bbf by Victor Stinner (Miss Islington (bot)) in branch '3.6':
bpo-31770: Prevent a crash and refleaks when calling sqlite3.Cursor.\_\_init__() more than once (GH-3968) (bpo-4301)
https://github.com/python/cpython/commit/9684cf69e32ae442c7be54521073ac78557f3bbf

**Comment by vstinner** [comment]
Thank you Oren Milman for your bug report and the bug fix! I merged your PR and backported it Python 3.6 and 2.7.
